src/openassetio-python/bridge/include/openassetio/python/converter.hpp

```diff
@@ -0,0 +1,105 @@
+ * `shared_ptr`, ensuring it is kept alive. The reference will be
+ * released when the `PyObject` is destroyed.
+ *
+ * @note This function does not acquire the GIL.
+ *
+ * @warning A Python environment, with openassetio imported, must be
+ *          available in order to use this function.
+ *
+ * @throws std::invalid_argument if the input is null.
+ * @throws std::runtime_error if the cast fails.
+ *
+ * @param objectPtr An OpenAssetIO pointer type, (eg, @needsref
+ * ManagerPtr). The returned pointer takes shared ownership of this
+ * input \p objectPtr, and will keep the C++ instance alive until the
+ * `PyObject` is destroyed.
+ *
+ * @return A `PyObject` pointer to an object of the Python API type
+ * associated with the C++ API object provided. The reference count of
+ * the `PyObject` will be incremented, and ownership of the reference
+ * transferred to the caller.
+ */
+template <typename T>
+OPENASSETIO_PYTHON_BRIDGE_EXPORT PyObject* castToPyObject(const T& objectPtr);
+
+/**
+ * Casts a Python object to the equivalent C++ API object.
+ *
+ * This template is explicitly instantiated to only the OpenAssetIO
+ * types, and is not intended to be a generic converter.
+ *
+ * The purpose of this function is to provide a Python/C++ conversion
+ * without requiring/exposing a particular CPython binding library
+ * implementation.
+ *
+ * The returned `shared_ptr` owns a hidden reference to the input
+ * `PyObject`, ensuring it is kept alive. The reference will be released
+ * when the `shared_ptr` is destroyed.
+ *
+ * Specifically, this function will increment the reference count of the
+ * provided Python object for the lifetime of the returned C++ object.
+ * When the returned C++ object falls out of scope or is otherwise
+ * cleaned up, the Python object will have its reference count reduced
+ * by one, potentially invoking cleanup.
+ *
+ * Using this function requires specifying the template argument of
+ * the C++ API type equivalent to the type of the object referred to by
+ * the \p pyObject pointer.
+ *
+ * @code{.cpp}
+ * ManagerPtr manager = castFromPyObject<Manager>(pyManager);
+ * @endcode
+ *
+ *
+ * If the types of the template argument and the \p pyObject are not
+ * equivalent, an exception will be thrown due to inability to perform
+ * the cast.
+ *
+ * @note This function does not acquire the GIL.
+ *
+ * @warning A Python environment, with openassetio imported, must be
+ *          available in order to use this function.
+ *
+ * @throws std::invalid_argument if the function fails due to inability
+ * to cast between types, or if the input is null.
+ *
+ * @param pyObject A `PyObject` pointer to a Python object that must be
+ * of equivalent type to the template argument.
+ *
+ * @return An OpenAssetIO pointer to a C++ API object cast from the
+ * provided \p pyObject. The lifetime of the PyObject will be extended
+ * to at least the lifetime of the returned C++ object.
```

Comment:
This seems contray to a prior suggestion.

In that, when referring to a `PyObject*`, we call it `Python Object`
So, when referring to a `std::shared_Ptr<T>`, should we not call it a `C++ Object`?

Reply:
For me it was just that "C++ object" suggests that it might be tied to the specific instance, whereas saying `shared_ptr` hopefully conveys that the lifetime is based on shared pointer semantics (ie, including all subsequent copies).